[BUG] @mastra/lance vector search not working

### Describe the Bug

I am trying to use `@mastra/lance` to integrate it with `Vector` searching, but for some reason is not working correctly; I am getting the following error:

```


Error calling handler HTTPException [Error]: Failed to execute query stream: GenericFailure, lance error: LanceError(Schema): Schema error: No field named metadata_resource_id. Valid fields are id, vector, _rowid, _rowaddr, _row_last_updated_at_version, _row_created_at_version, _rowoffset, _distance., /Users/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/lance-0.39.0/src/dataset/scanner.rs:260:37  Caused by: LanceError(Schema): Schema error: No field named metadata_resource_id. Valid fields are id, vector, _rowid, _rowaddr, _row_last_updated_at_version, _row_created_at_version, _rowoffset, _distance., /Users/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/lance-0.39.0/src/dataset/scanner.rs:260:37
    at _LanceVectorStore.query (file:///Users/x/Laboratory/x/mcp/node_modules/@mastra/lance/dist/index.js:2440:13)
    at async file:///Users/x/Laboratory/x/mcp/node_modules/@mastra/memory/dist/index.js:14662:16
    at async Promise.all (index 0)
    at async Memory.recall (file:///Users/x/Laboratory/x/mcp/node_modules/@mastra/memory/dist/index.js:14654:7)
    at async Object.handler (file:///Users/x/Laboratory/x/mcp/.mastra/output/index.mjs:6868:22)
    at async file:///Users/x/Laboratory/x/mcp/.mastra/output/index.mjs:52667:26
    at async dispatch (file:///Users/x/Laboratory/x/mcp/.mastra/output/index.mjs:48090:17)
    at async dispatch (file:///Users/x/Laboratory/x/mcp/.mastra/output/index.mjs:48090:17)
    at async handler (file:///Users/x/Laboratory/x/mcp/.mastra/output/index.mjs:2541:9)
    at async dispatch (file:///Users/x/Laboratory/x/mcp/.mastra/output/index.mjs:48090:17) {
  res: undefined,
  status: 500,
  [cause]: [Error: Failed to execute query stream: GenericFailure, lance error: LanceError(Schema): Schema error: No field named metadata_resource_id. Valid fields are id, vector, _rowid, _rowaddr, _row_last_updated_at_version, _row_created_at_version, _rowoffset, _distance., /Users/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/lance-0.39.0/src/dataset/scanner.rs:260:37  Caused by: LanceError(Schema): Schema error: No field named metadata_resource_id. Valid fields are id, vector, _rowid, _rowaddr, _row_last_updated_at_version, _row_created_at_version, _rowoffset, _distance., /Users/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/lance-0.39.0/src/dataset/scanner.rs:260:37] {
    code: 'GenericFailure'
  }
}
```

Memory Configuration is: 

```
  memory: new Memory({
    vector: await LanceVectorStore.create(getEnvOrThrow("SV_MCP_STORAGE")),
    embedder: fastembed,
    options: {
      semanticRecall: {
        topK: 3, // Retrieve 3 most similar messages
        messageRange: 2, // Include 2 messages before and after each match
        scope: "resource", // Search across all threads for this user (default setting if omitted)
      },
      workingMemory: {
        enabled: true,
        scope: "thread",
      },
    },
  }),
```

### Steps To Reproduce

1. Create project
2. Install lance and fastembed
3. Add lance/fastembed to Memory configuration

### Link to Minimal Reproducible Example

N/A

### Expected Behavior

Should expect vector search results

### Environment Information

```block
N/A
```

### Verification

- [x] I have searched the existing issues to make sure this is not a duplicate
- [x] I have included sufficient information for the team to reproduce and understand the issue

--- a/.changeset/sparkly-sides-chew.md
+++ b/.changeset/sparkly-sides-chew.md
@@ -1,0 +1,5 @@
+---
+'@mastra/lance': patch
+---
+
+Fixed Lance vector store failing when used with Memory's semantic recall. Queries on empty tables with metadata filters (e.g. resource_id, thread_id) now return empty results instead of throwing a schema error. Also fixed metadata round-trip corruption where flat underscore keys like resource_id were incorrectly reconstructed as nested objects.

--- a/stores/lance/src/vector/index.test.ts
+++ b/stores/lance/src/vector/index.test.ts
@@ -1150,7 +1150,9 @@ describe('Lance vector store tests', () => {
         expect(res).toHaveLength(1);
         expect(res[0].id).toBe(ids[0]);
         expect(res[0].metadata?.name).to.equal('test2');
-        expect(res[0].metadata?.details?.text).to.equal('test2');
+        // Metadata is flat when _metadata_json is not available (table created via createTable
+        // without _metadata_json column). Nested key 'details.text' becomes 'details_text'.
+        expect(res[0].metadata?.details_text).to.equal('test2');
       });
 
       it('should not throw error when filter is not provided', async () => {
@@ -1512,7 +1514,7 @@ describe('Lance vector store tests', () => {
       });
 
       it('should query with nested comparison and pattern matching', async () => {
-        const testTableName = 'test-nested-table';
+        const nestedTableName = 'test-nested-table-' + Date.now();
 
         const testVectors = [
           [0.7, 0.7, 0.7],
@@ -1521,7 +1523,7 @@ describe('Lance vector store tests', () => {
 
         const ids = await vectorDB.upsert({
           indexName: testTableIndexColumn,
-          tableName: testTableName,
+          tableName: nestedTableName,
           vectors: testVectors,
           metadata: [
             {
@@ -1545,7 +1547,7 @@ describe('Lance vector store tests', () => {
 
         const res = await vectorDB.query({
           indexName: testTableIndexColumn,
-          tableName: testTableName,
+          tableName: nestedTableName,
           queryVector: [0.75, 0.75, 0.75],
           topK: 10,
           includeAllColumns: true,
@@ -1560,6 +1562,9 @@ describe('Lance vector store tests', () => {
         const usernamesFound = res.map(item => item.metadata?.profile?.username);
         expect(usernamesFound).toContain('john_doe');
         expect(usernamesFound).toContain('jane_smith');
+
+        // Cleanup
+        await vectorDB.deleteTable(nestedTableName);
       });
     });
 
@@ -1937,6 +1942,42 @@ describe('Lance vector store tests', () => {
         await vectorDB.deleteTable(indexName);
       });
 
+      it('should handle empty recall with filter (query before any upsert, issue #12500)', async () => {
+        const indexName = 'memory_flow_empty_filter_' + Date.now();
+
+        // Memory flow: createIndex first (creates table with {id, vector, _metadata_json} only)
+        await vectorDB.createIndex({
+          indexName,
+          dimension: 3,
+          metric: 'cosine',
+        });
+
+        // Memory.recall() queries with resource_id filter BEFORE any saveMessages/upsert.
+        // The table has no metadata_resource_id column yet.
+        // This should return empty results, not throw a schema error.
+        const resourceResults = await vectorDB.query({
+          indexName,
+          queryVector: [0.1, 0.2, 0.3],
+          topK: 5,
+          filter: { resource_id: 'user-123' },
+        });
+
+        expect(resourceResults).toEqual([]);
+
+        // Same for thread_id filter
+        const threadResults = await vectorDB.query({
+          indexName,
+          queryVector: [0.1, 0.2, 0.3],
+          topK: 5,
+          filter: { thread_id: 'thread-456' },
+        });
+
+        expect(threadResults).toEqual([]);
+
+        // Cleanup
+        await vectorDB.deleteTable(indexName);
+      });
+
       it('should handle save then recall flow', async () => {
         const indexName = 'memory_flow_save_recall_' + Date.now();
 
@@ -1971,6 +2012,149 @@ describe('Lance vector store tests', () => {
 
         expect(results.length).toBe(2);
         expect(results[0].id).toBeDefined();
+
+        // Cleanup
+        await vectorDB.deleteTable(indexName);
+      });
+    });
+
+    describe('metadata round-trip with underscore keys (issue #12500)', () => {
+      it('should preserve flat underscore keys through upsert and query', async () => {
+        const indexName = 'roundtrip_flat_keys_' + Date.now();
+
+        // 1. createIndex (Memory does this first)
+        await vectorDB.createIndex({
+          indexName,
+          dimension: 3,
+          metric: 'cosine',
+        });
+
+        // 2. upsert with flat underscore keys (exactly what Memory does)
+        await vectorDB.upsert({
+          indexName,
+          vectors: [
+            [0.1, 0.2, 0.3],
+            [0.4, 0.5, 0.6],
+          ],
+          metadata: [
+            { message_id: 'msg1', thread_id: 'thread1', resource_id: 'user1' },
+            { message_id: 'msg2', thread_id: 'thread1', resource_id: 'user1' },
+          ],
+        });
+
+        // 3. query without filter — metadata should round-trip correctly
+        const results = await vectorDB.query({
+          indexName,
+          queryVector: [0.1, 0.2, 0.3],
+          topK: 2,
+        });
+
+        expect(results.length).toBe(2);
+
+        // BUG: unflattenObject splits on ALL underscores, so
+        // { resource_id: 'user1' } becomes { resource: { id: 'user1' } }
+        // This asserts the CORRECT behavior:
+        expect(results[0].metadata).toHaveProperty('resource_id');
+        expect(results[0].metadata).toHaveProperty('thread_id');
+        expect(results[0].metadata).toHaveProperty('message_id');
+        // These should NOT exist (but they do due to the bug):
+        expect(results[0].metadata).not.toHaveProperty('resource');
+        expect(results[0].metadata).not.toHaveProperty('thread');
+        expect(results[0].metadata).not.toHaveProperty('message');
+
+        // Cleanup
+        await vectorDB.deleteTable(indexName);
+      });
+
+      it('should filter by flat underscore keys in Memory-like recall flow', async () => {
+        const indexName = 'recall_filter_underscore_' + Date.now();
+
+        // 1. createIndex
+        await vectorDB.createIndex({
+          indexName,
+          dimension: 3,
+          metric: 'cosine',
+        });
+
+        // 2. upsert with Memory-style metadata
+        await vectorDB.upsert({
+          indexName,
+          vectors: [
+            [0.1, 0.2, 0.3],
+            [0.4, 0.5, 0.6],
+            [0.7, 0.8, 0.9],
+          ],
+          metadata: [
+            { message_id: 'msg1', thread_id: 'thread1', resource_id: 'user1' },
+            { message_id: 'msg2', thread_id: 'thread1', resource_id: 'user1' },
+            { message_id: 'msg3', thread_id: 'thread2', resource_id: 'user2' },
+          ],
+        });
+
+        // 3. query with resource_id filter (exactly what Memory.recall does with scope: "resource")
+        // This is the exact call path that triggers the reported error
+        const results = await vectorDB.query({
+          indexName,
+          queryVector: [0.1, 0.2, 0.3],
+          topK: 5,
+          filter: { resource_id: 'user1' },
+        });
+
+        // Should return only the 2 results for user1
+        expect(results.length).toBe(2);
+        for (const result of results) {
+          expect(result.metadata?.resource_id).toBe('user1');
+        }
+
+        // Also test thread_id filter (Memory.recall with scope: "thread")
+        const threadResults = await vectorDB.query({
+          indexName,
+          queryVector: [0.1, 0.2, 0.3],
+          topK: 5,
+          filter: { thread_id: 'thread2' },
+        });
+
+        expect(threadResults.length).toBe(1);
+        expect(threadResults[0].metadata?.thread_id).toBe('thread2');
+
+        // Cleanup
+        await vectorDB.deleteTable(indexName);
+      });
+
+      it('should not corrupt metadata when keys contain underscores mixed with nested objects', async () => {
+        const indexName = 'mixed_underscore_nested_' + Date.now();
+
+        // 1. createIndex
+        await vectorDB.createIndex({
+          indexName,
+          dimension: 3,
+          metric: 'cosine',
+        });
+
+        // 2. upsert with BOTH flat underscore keys and nested objects
+        await vectorDB.upsert({
+          indexName,
+          vectors: [[0.1, 0.2, 0.3]],
+          metadata: [
+            {
+              resource_id: 'user1',
+              details: { context: 'test' },
+            },
+          ],
+        });
+
+        // 3. query and check metadata structure
+        const results = await vectorDB.query({
+          indexName,
+          queryVector: [0.1, 0.2, 0.3],
+          topK: 1,
+        });
+
+        expect(results.length).toBe(1);
+        // resource_id should remain flat
+        expect(results[0].metadata?.resource_id).toBe('user1');
+        // details.context should be preserved as nested
+        expect(results[0].metadata?.details).toEqual({ context: 'test' });
 
         // Cleanup
         await vectorDB.deleteTable(indexName);

--- a/stores/lance/src/vector/index.ts
+++ b/stores/lance/src/vector/index.ts
@@ -155,6 +155,22 @@ export class LanceVectorStore extends MastraVector<LanceVectorFilter> {
       if (filter && Object.keys(filter).length > 0) {
         const whereClause = this.filterTranslator(filter);
         this.logger.debug(`Where clause generated: ${whereClause}`);
+
+        // Validate that filter columns exist in the table schema.
+        // When Memory.recall() runs before any saveMessages(), the table has no metadata columns yet.
+        // Filtering on non-existent columns would throw a LanceDB schema error.
+        const schema = await table.schema();
+        const schemaColumns = new Set(schema.fields.map((f: any) => f.name));
+        const filterColumns = this.extractFilterColumns(whereClause);
+        const missingColumns = filterColumns.filter(col => !schemaColumns.has(col));
+
+        if (missingColumns.length > 0) {
+          this.logger.debug(
+            `Filter references columns not in schema: ${missingColumns.join(', ')}. Returning empty results.`,
+          );
+          return [];
+        }
+
         query = query.where(whereClause);
       }
 
@@ -173,23 +189,19 @@ export class LanceVectorStore extends MastraVector<LanceVectorFilter> {
       const results = await query.toArray();
 
       return results.map(result => {
-        // Collect all metadata_ prefixed fields
-        const flatMetadata: Record<string, any> = {};
+        let metadata: Record<string, any> = {};
 
-        // Get all keys from the result object
-        Object.keys(result).forEach(key => {
-          // Skip reserved keys (id, score, and the vector column)
-          if (key !== 'id' && key !== 'score' && key !== 'vector' && key !== '_distance') {
-            if (key.startsWith('metadata_')) {
-              // Remove the prefix and add to flat metadata
-              const metadataKey = key.substring('metadata_'.length);
-              flatMetadata[metadataKey] = result[key];
-            }
+        if (result._metadata_json) {
+          // New data: parse the lossless JSON column
+          try {
+            metadata = JSON.parse(result._metadata_json);
+          } catch {
+            metadata = this.extractFlatMetadata(result);
           }
-        });
-
-        // Reconstruct nested metadata object
-        const metadata = this.unflattenObject(flatMetadata);
+        } else {
+          // Legacy data or empty metadata: extract flat metadata keys as-is
+          metadata = this.extractFlatMetadata(result);
+        }
 
         return {
           id: String(result.id || ''),
@@ -321,6 +333,10 @@ export class LanceVectorStore extends MastraVector<LanceVectorFilter> {
           Object.entries(flattenedMetadata).forEach(([key, value]) => {
             rowData[key] = value;
           });
+          // Store lossless JSON representation for round-trip fidelity
+          rowData['_metadata_json'] = JSON.stringify(metadataItem);
+        } else {
+          rowData['_metadata_json'] = '';
         }
 
         return rowData;
@@ -572,7 +588,9 @@ export class LanceVectorStore extends MastraVector<LanceVectorFilter> {
         );
 
         const initVector = new Array(dimension).fill(0);
-        table = await this.lanceClient.createTable(resolvedTableName, [{ id: '__init__', vector: initVector }]);
+        table = await this.lanceClient.createTable(resolvedTableName, [
+          { id: '__init__', vector: initVector, _metadata_json: '' },
+        ]);
         try {
           await table.delete("id = '__init__'");
         } catch (deleteError) {
@@ -985,9 +1003,24 @@ export class LanceVectorStore extends MastraVector<LanceVectorFilter> {
 
               // Apply metadata updates if provided
               if (update.metadata) {
+                // Update flattened columns for filter compatibility
                 Object.entries(update.metadata).forEach(([key, value]) => {
                   rowData[`metadata_${key}`] = value;
                 });
+
+                // Keep _metadata_json in sync if the table has it
+                const hasMetadataJson = schema.fields.some((f: any) => f.name === '_metadata_json');
+                if (hasMetadataJson) {
+                  let existingMetadata: Record<string, any> = {};
+                  if (record._metadata_json) {
+                    try {
+                      existingMetadata = JSON.parse(record._metadata_json);
+                    } catch {
+                      // If existing JSON is corrupt, start fresh with updates only
+                    }
+                  }
+                  rowData['_metadata_json'] = JSON.stringify({ ...existingMetadata, ...update.metadata });
+                }
               }
 
               return rowData;
@@ -1097,40 +1130,29 @@ export class LanceVectorStore extends MastraVector<LanceVectorFilter> {
   }
 
   /**
-   * Converts a flattened object with keys using underscore notation back to a nested object.
-   * Example: { name: 'test', details_text: 'test' } → { name: 'test', details: { text: 'test' } }
+   * Extracts column names referenced in a SQL WHERE clause.
+   * Identifies metadata_* prefixed identifiers used in filter conditions.
    */
-  private unflattenObject(obj: Record<string, any>): Record<string, any> {
-    const result: Record<string, any> = {};
+  private extractFilterColumns(whereClause: string): string[] {
+    // Match identifiers that look like column names (word characters, starting with metadata_)
+    const matches = whereClause.match(/metadata_\w+/g);
+    return matches ? [...new Set(matches)] : [];
+  }
 
-    Object.keys(obj).forEach(key => {
-      const value = obj[key];
-      const parts = key.split('_');
-
-      // Start with the result object
-      let current = result;
-
-      // Process all parts except the last one
-      for (let i = 0; i < parts.length - 1; i++) {
-        const part = parts[i];
-        // Skip empty parts
-        if (!part) continue;
-
-        // Create nested object if it doesn't exist
-        if (!current[part] || typeof current[part] !== 'object') {
-          current[part] = {};
+  /**
+   * Extracts metadata from flattened column names (legacy data without _metadata_json).
+   * Returns keys as-is after stripping the 'metadata_' prefix, without any unflattening.
+   */
+  private extractFlatMetadata(result: Record<string, any>): Record<string, any> {
+    const metadata: Record<string, any> = {};
+    Object.keys(result).forEach(key => {
+      if (key !== 'id' && key !== 'score' && key !== 'vector' && key !== '_distance' && key !== '_metadata_json') {
+        if (key.startsWith('metadata_')) {
+          metadata[key.substring('metadata_'.length)] = result[key];
         }
-        current = current[part];
-      }
-
-      // Set the value at the last part
-      const lastPart = parts[parts.length - 1];
-      if (lastPart) {
-        current[lastPart] = value;
       }
     });
-
-    return result;
+    return metadata;
   }
 
   async deleteVectors({ indexName, filter, ids }: DeleteVectorsParams<LanceVectorFilter>): Promise<void> {

